Nwbinspector (#4)
catalystneuro/2023-ecosystem-paper: README.md

## Results
### NWBInspector

> It can be time-consuming and error-prone to manually inspect all of the NWB files in a dataset, especially as datasets grow very large. DANDI datasets frequently occupy hundreds of GBs to TBs of disk space, spread across thousands of individual NWB files [cite DANDI]. This challenge is compounded for new users who may lack familiarity with NWB best practices, further hampering the effectiveness of manual inspection. To address this problem, we have developed _NWBInspector_, a Python-based package that assesses the quality of NWB files based on their compliance with established best practices.

Reviewer: ```suggestion
It can be time-consuming and error-prone to manually inspect all of the NWB files in a dataset, especially as datasets grow very large. DANDI datasets frequently occupy hundreds of GBs to TBs of disk space, spread across thousands of individual NWB files [cite DANDI]. This challenge is compounded for new users who may lack familiarity with NWB best practices, further hampering the effectiveness of manual inspection. To address this problem, we have developed _NWB Inspector_, a Python-based package that assesses the quality of NWB files based on their compliance with established best practices.
```

After installation, _NWBInspector_ can be used very easily from the command line. Simply run
```bash
    nwbinspector path/to/my/data.nwb
```
to inspect a single NWB file or
```bash

>         nwbinspector path/to/my/data/dir/

Reviewer: ```suggestion
    nwbinspector path/to/my/data/dir/
```

```
to inspect all the NWB files in a given directory.

> This command will produce a report of all the issues flagged by _NWBInspector_ directly in the terminal, ordered by importance as shown below.

Reviewer: ```suggestion
This command will produce a report of all the issues flagged by _NWB Inspector_ directly in the terminal, ordered by importance as shown below.
```

[Example output]

> Unlike the _PyNWB_ validator, which checks for strict schema compliance, _NWBInspector_ only serves as an automatic data review aid. This means that, in an effort to catch as many potential mistakes as possible, _NWBInspector_ may raise warnings for intentional deviations from standard best practices. Conversely, if a user makes a mistake that does not violate any established best practices and it passes the _PyNWB_ validator, _NWBInspector_ might not catch that mistake. This structure leads to a trade-off between comprehensive inspection, which catches as many potential mistakes as possible, and an unobtrusive inspection process that only highlights truly problematic choices rather than intentional but unorthodox decisions.

Reviewer: ```suggestion
Unlike the _PyNWB_ validator, which checks for strict schema compliance, _NWBInspector_ only serves as an automatic data review aid. This means that, in an effort to catch as many potential mistakes as possible, _NWB Inspector_ may raise warnings for intentional deviations from standard best practices. Conversely, if a user makes a mistake that does not violate any established best practices and it passes the _PyNWB_ validator, _NWBInspector_ might not catch that mistake. This structure leads to a trade-off between comprehensive inspection, which catches as many potential mistakes as possible, and an unobtrusive inspection process that only highlights truly problematic choices rather than intentional but unorthodox decisions.
```

> Unlike the _PyNWB_ validator, which checks for strict schema compliance, _NWBInspector_ only serves as an automatic data review aid. This means that, in an effort to catch as many potential mistakes as possible, _NWB Inspector_ may raise warnings for intentional deviations from standard best practices. Conversely, if a user makes a mistake that does not violate any established best practices and it passes the _PyNWB_ validator, _NWBInspector_ might not catch that mistake. This structure leads to a trade-off between comprehensive inspection, which catches as many potential mistakes as possible, and an unobtrusive inspection process that only highlights truly problematic choices rather than intentional but unorthodox decisions.

Reviewer: ```suggestion
Unlike the _PyNWB_ validator, which checks for strict schema compliance, _NWB Inspector_ only serves as an automatic data review aid. This means that, in an effort to catch as many potential mistakes as possible, _NWB Inspector_ may raise warnings for intentional deviations from standard best practices. Conversely, if a user makes a mistake that does not violate any established best practices and it passes the _PyNWB_ validator, _NWB Inspector_ might not catch that mistake. This structure leads to a trade-off between comprehensive inspection, which catches as many potential mistakes as possible, and an unobtrusive inspection process that only highlights truly problematic choices rather than intentional but unorthodox decisions.
```

> Our solution to this trade-off involves splitting the results into three tiers of warnings/errors. The first tier is for best practice violations, which are clear violations of known best practices, such as negative spike times. The second tier is for critical issues, which are potentially critical mistakes, such as zero-rate timestamps. The third tier is for best practice suggestions, which are recommendations for improving the NWB file, such as adding an informative description. This approach allows our inspection process to be as comprehensive as possible while minimizing disruptions from ignoring specific best practice suggestions.

Reviewer: Critical issues come before violations

Critical issues are essentially invalidations that are too complicated to define in the schema language or even in PyNWB logic. They can also be heuristical checks of invalid fields that can't be identified deterministically.

> Power users of NWBInspector often seek more direct control of the inspection process and its outputs. To enable more flexibility when using _NWBInspector_, we provide a simple Python API with all the same functionality as the CLI, along with some more advanced features. Just like the CLI, the Python API can be employed to inspect individual NWB files and entire directories of NWB files with the `nwbinspector.inspect_nwbfile()` function and `nwbinspector.inspect_all()` function respectively. Additionally, the Python API allows users to inspect NWB files directly in memory with the `nwbinspector.inspect_nwbfile_object()` function. Since both the `inspect_nwbfile` and the `inspect_all` functions return generators, Python users can employ these functions in more sophisticated ways, such as implementing early stopping if a particular message is raised. These features empower advanced users to fully leverage _NWBInspector_.

Reviewer: ```suggestion
Power users of NWB Inspector often seek more direct control of the inspection process and its outputs. To enable more flexibility when using _NWB Inspector_, we provide a simple Python API with all the same functionality as the CLI, along with some more advanced features. Just like the CLI, the Python API can be employed to inspect individual NWB files and entire directories of NWB files with the `nwbinspector.inspect_nwbfile()` function and `nwbinspector.inspect_all()` function respectively. Additionally, the Python API allows users to inspect NWB files directly in memory with the `nwbinspector.inspect_nwbfile_object()` function. Since both the `inspect_nwbfile` and the `inspect_all` functions return generators, Python users can employ these functions in more sophisticated ways, such as implementing early stopping if a particular message is raised. These features empower advanced users to fully leverage _NWB Inspector_.
```

> Some organizations have their members use NWBInspector, but they may need to adjust the importance level of some checks to meet their requirements. For instance, DANDI requires users to specify the sex and age of their subjects, so those checks should be elevated to importance level "critical" instead of "best practice suggestions." To accommodate this need, NWBInspector provides a configuration option where the importance level of individual checks can be changed with a simple YAML file. Similarly, adding new checks is as simple as defining the check function in some module with a `register_check` decorator, and then specifying the module name with the `--modules` flag in the CLI. For example, in the module defining some new check:

Reviewer: ```suggestion
Some organizations have their members use NWB Inspector, but they may need to adjust the importance level of some checks to meet their requirements. For instance, DANDI requires users to specify the sex and age of their subjects, so those checks should be elevated to importance level "critical" instead of "best practice suggestions." To accommodate this need, NWB Inspector provides a configuration option where the importance level of individual checks can be changed with a simple YAML file. Similarly, adding new checks is as simple as defining the check function in some module with a `register_check` decorator, and then specifying the module name with the `--modules` flag in the CLI. For example, in the module defining some new check:
```

```python
from nwbinspector.register_checks import available_checks, register_check, Importance

@register_check(importance=Importance.SOME_IMPORTANCE_LEVEL, neurodata_type=some_neurodata_type)
def check_personal_practice(...):
    ...
```
and then in the command line, include that user-defined check,
```bash

>     nwbinspector --modules path/to/module.py path/to/my/data.nwb

Reviewer: ```suggestion
    nwbinspector --modules name_of_module path/to/my/data.nwb
```

```

> NWB has various Neurodata objects that all require appropriate checks, which are often the same or very similar across different objects. For instance, ROIResponseSeries, ElectricalSeries, and OptogeneticSeries all need to verify that their timestamps are non-negative and ascending. To avoid unnecessary duplication of checks, we leverage the inheritance of Neurodata objects in NWB and apply checks to all the children of a given Neurodata object. In this case, ROIResponseSeries, ElectricalSeries, and OptogeneticSeries are all descendants of TimeSeries. Thus, when we define `check_timestamps_ascending` for TimeSeries, it is automatically applied to all these objects. This approach allows NWBInspector to provide a minimal set of checks that leverages the inheritance structure of NWB to assess all the relevant objects.

Reviewer: ```suggestion
NWB has various neurodata objects that all require appropriate checks, which are often the same or very similar across different objects. For instance, ROIResponseSeries, ElectricalSeries, and OptogeneticSeries all need to verify that their timestamps are non-negative and ascending. To avoid unnecessary duplication of checks, we leverage the inheritance of Neurodata objects in NWB and apply checks to all the children of a given Neurodata object. In this case, ROIResponseSeries, ElectricalSeries, and OptogeneticSeries are all descendants of TimeSeries. Thus, when we define `check_timestamps_ascending` for TimeSeries, it is automatically applied to all these objects. This approach allows NWB Inspector to provide a minimal set of checks that leverages the inheritance structure of NWB to assess all the relevant objects.
```

> In summary, _NWBInspector_ streamlines NWB file inspection with a user-friendly command-line interface, tiered warnings, and a flexible Python API. It strikes a balance between comprehensive inspection and unobtrusiveness, catering to both new and power users. The tool adapts to organizational needs, allowing customization of checks and easy incorporation of new ones. Leveraging the inheritance structure of NWB Neurodata objects, _NWBInspector_ ensures a minimal yet effective set of checks for comprehensive assessment. Overall, it stands as a versatile and efficient solution for curation of NWB datasets.

Reviewer: ```suggestion
In summary, _NWB Inspector_ streamlines NWB file inspection with a user-friendly command-line interface, tiered warnings, and a flexible Python API. It strikes a balance between comprehensive inspection and unobtrusiveness, catering to both new and power users. The tool adapts to organizational needs, allowing customization of checks and easy incorporation of new ones. Leveraging the inheritance structure of NWB Neurodata objects, _NWB Inspector_ ensures a minimal yet effective set of checks for comprehensive assessment. Overall, it stands as a versatile and efficient solution for curation of NWB datasets.
```
